src/replay/consensus.zig

```diff
@@ -443,8 +460,68 @@ fn resetFork(
     _ = last_reset_bank_descendants;
 }
 
+fn computeBankStats(
+    allocator: std.mem.Allocator,
+    my_vote_pubkey: Pubkey,
+    ancestors: *const std.AutoArrayHashMapUnmanaged(u64, SortedSet(u64)),
+    slot_tracker: *SlotTracker,
+    epoch_stakes: *const EpochStakesMap,
+    epoch_schedule: *const EpochSchedule,
+    progress: *ProgressMap,
+    fork_choice: *ForkChoice,
+    latest_validator_votes: *LatestValidatorVotesForFrozenBanks,
+) ![]Slot {
+    var new_stats = std.ArrayListUnmanaged(Slot).empty;
+    errdefer new_stats.deinit(allocator);
+    var frozen_slots = try slot_tracker.frozenSlots(allocator);
+    defer frozen_slots.deinit(allocator);
+    // TODO agave sorts this by the slot first. Is this needed for the implementation to be correct?
+    // If not, then we can avoid sorting here which may be verbose given frozen_slots is a map.
+    for (frozen_slots.keys()) |slot| {
+        const epoch = epoch_schedule.getEpoch(slot);
+        const epoch_stake = epoch_stakes.get(epoch) orelse return error.MissingEpochStake;
+        const fork_stat = progress.getForkStats(slot) orelse return error.MissingSlot;
+        if (!fork_stat.computed) {
+            // TODO Self::adopt_on_chain_tower_if_behind
+            // Gather voting information from all vote accounts to understand the current consensus state.
+            const computed_bank_state = try collectVoteLockouts(
+                allocator,
+                .noop,
+                &my_vote_pubkey,
+                slot,
+                &epoch_stake.stakes.vote_accounts.vote_accounts,
+                ancestors,
+                progress,
+                latest_validator_votes,
+            );
+            try fork_choice.computeBankStats(
+                allocator,
+                epoch_stakes,
+                epoch_schedule,
+                latest_validator_votes,
+            );
+            const fork_stats = progress.getForkStats(slot) orelse return error.MissingForkStats;
+            fork_stats.fork_stake = computed_bank_state.fork_stake;
+            fork_stats.total_stake = computed_bank_state.total_stake;
+            fork_stats.voted_stakes = computed_bank_state.voted_stakes;
+            fork_stats.lockout_intervals =
+                try computed_bank_state.lockout_intervals.clone(allocator);
```

Comment:
You don't need to clone this since computed_bank_state is being discarded in this loop. You should be able to just transfer ownership here, similar to voted_stakes.

Actually, it's strange that this does not show up as a leak in your tests, since computed_bank_state is never deinited. I just ran all of them tests for computeBankStats, and it looks like lockout_intervals is always empty in every test. What do you think about modifying one of the tests to make it non-empty? Maybe that would improve test coverage.

Reply:
This ensure at least one lockout [918d2ed](https://github.com/Syndica/sig/pull/812/commits/918d2edb3774ef1a809f1713cb537c6a46ab28a3)

Revealed a leak in the tests that was handled in the deinit in the test fixture.